### ISSUE TITLE
Generate a suppressions.xml for SuppressionFilter file from the errors list by CLI command

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#5983]
originally was reports as #102

Link to patch: https://sourceforge.net/p/checkstyle/patches/_discuss/thread/1d416f58/b6be/attachment/mychanges.patch

Explanation from patch author:
http://juliangamble.com/blog/2012/03/10/running-checkstyle-at-full-throttle-on-a-legacy-codebase/

Few thoughts about reasons and usage of this patch:
http://roman-ivanov.blogspot.com/2014/01/checkstyle-suppress-file-generator.html

**Comment by rnveach** [comment]
1st link doesn't work anymore.

Basically the patch made a AuditListener that took violations and printed a suppression file instead.

This is very similar to `XpathFileGeneratorAuditListener`, so much so all that is needed is to not print the xpath and you can use the same code.

@romani Why don't we just merge the 2 into 1 and print suppressions for non-xpath events.
https://github.com/checkstyle/checkstyle/blob/592bda25a29b119ef1dfccc6fae2760e3057cc39/src/main/java/com/puppycrawl/tools/checkstyle/XpathFileGeneratorAuditListener.java#L94-L95

Instead of printing nothing if we can't get the xpath, just print everything except query.

Since this AuditListener is internal to us, I think we should also rename it since it won't deal specifically with xpath anymore.

**Comment by romani** [comment]
A lot time passed from original request.

Your idea is Interesting, but it might be confusing for user, as he requested an xpath but will give him smth else. We cli option should be used, but code internally can be heavily reused and refactored

**Comment by rnveach** [comment]
> it might be confusing for user, as he requested an xpath but will give him smth else

The point is to change everything to say "generate suppression" and not specifically say "xpath" to avoid any confusion. If its Java and supports xpath, then it will generate an xpath suppression, otherwise it will generate the normal suppression.

**Comment by romani** [comment]
>normal suppression.

Do we have such term ? We have multiple ways to suppress, xpath, file, line number

**Comment by romani** [comment]
I am ok with idea, I just want to understand a scope of fix in this issue

in issue description:
>All we need to do is to suppress certain Check (id or name) on whole file. It is not ideal but better than nothing.

you mentioned:
> Instead of printing nothing if we can't get the xpath, just print everything except query.

in CLI we have support only for xpath for now: https://checkstyle.org/cmdline.html#Command_line_usage
Generation of Suppression by whole file do make sense for big legacy projects.

If you want to discuss xpath suppression generations ... lets do this in separate issue.

**Comment by rnveach** [comment]
> I just want to understand a scope of fix in this issue

Ok let me clarify. I see where some confusion is coming from.

I am talking about changing `--generate-xpath-suppression` option to just `--generate-suppressions`. We will change `XpathFileGeneratorAuditListener` to just `SuppressionFileGeneratorAuditListener`. The only change in the class is it will print non-xpath suppressions if an xpath suppression can't be generated.

So class will print `<suppress-xpath files="..."  checks="..." query="..."/>` for xpaths and `<suppress files="..."  checks="..."/>` for non-xpaths.

No other changes.

> -s line:column - print xpath suppression

This won't change at all. It will still be xpath only.

**Comment by romani** [comment]
I do not want to mix type of  generated output.
I see big use in generation of suppression for particular check on whole file.
Few  other modes that might be interesting:
- For check  with message
- for check and range of lines
- for check and xpath 

So user can choose a proper strategy of suppression. So it might result in several Cli modes.

**Comment by Fernal73** [comment]
Is there any way other than redirection to save the xpath file generated with the -g option? Is there going to be a suppressions.xml generated instead? Any time soon?

**Comment by romani** [comment]
Unlikely we would generate a file, it is easy to redirect output to file to get required result

**Comment by Fernal73** [comment]
Agreed. I see that. There's just one line to be deleted. What's the best way to use this assuming I'm using a global checkstyle configuration and wish to customise suppressions for individual projects?
What's the best way to handle this?
I've downloaded the full checkstyles.xml that runs almost all your modules and also have the google_checks.xml file on which there are a few modifications of my own.
Specifying the xpath generated file threw an exception claiming that xpath expression was unexpected.

My project resides at https://github.com/Fernal73/LearnJava and the specific folder you'd wish to run checkstyle against is Facade2.

I'm currently running the batch files in the root folder to run checkstyle.

Thank you for your time.

**Comment by Fernal73** [comment]
The generated xpath file is under folder Facade2 and  named
suppressions-xpath.xml. You can uncomment the Suppression-Xpath-Filter
module in google_checks.xml to recreate the parser exception.

The generated xpath file is quite large and lists out every checkstyle
error. That might be useful to retain if your code is a living entity and
you need to catch any errors in new methods or code added.


 I prefer the format of the suppressions.xml but running that module
against my current suppressions.xml file in Facade2 made the process hang.
Is there something I'm missing?

The SuppressionXPathSingleFilter works wonderfully well but I'm not
comfortable having that in the global configuration.

Can something like that be dynamically but temporarily inserted into the
global configuration? A merged configuration?

Or is there some way to get the suppressions.xml and suppressions-xpath.xml
working as expected?

Regards,
Linus.

**Comment by romani** [comment]
@Fernal73 ,

please move your issue details to https://groups.google.com/forum/#!forum/checkstyle , lets not pollute in this issue. I will ask some questions in mailslist to guide you on how to investiagate your problem.

**Comment by Fernal73** [comment]
OK. Will do.

I'll need to investigate this further myself. Tricky regex, perhaps.

The original checkstyle.xml with suppressions.xml and suppression-path.xml
in the root directory worked perfectly.

Maybe, I'm just too tired to think clearly.
Perhaps, I'll sleep on it.

Meanwhile, I've subscribed to the Checkstyle group.

The Checkstyle group doesn't show up in search listings when looking to
resolve issues.

Regards,
Linus.

**Comment by Harsh-Agrawal96** [comment]
We already have the `suppressions.xml` generation for `XpathSuppressFilter`
and now How `SuppressFilter` is differ from the `XpathSuppressFilter`, what changes need to do for done the generation of `suppression.xml` for `SuppressFilter`

**Comment by Harsh-Agrawal96** [comment]
I am working on this issue.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/pull/16916 is merged.

@Harsh-Agrawal96, thanks a lot for your contribution.
